### ISSUE TITLE
Infinite loop / corruption if copying 4 bytes or less with both source and target misaligned using memcpy

### ISSUE THREAD

**Issue by gwilymk** [felixjones/agbabi#6]
While integrating agbabi into agbrs, we found that running some of the larger projects would crash after some time, and narrowed it down to agbabi's implementation of memcpy.

Copying less than 4 bytes if both the source and target pointers are 1 byte misaligned seems to cause this issue (see this test here: https://github.com/agbrs/agb/blob/c9f1401ef3ab5e5c78b27016a8332691d05ca2af/agb/src/agbabi/mod.rs#L126)

Here is a modification I made which fixes the issue: https://github.com/agbrs/agb/blob/c9f1401ef3ab5e5c78b27016a8332691d05ca2af/agb/src/agbabi/memcpy.s#L32 (at least for all the test cases I've added). However, it isn't ideal when the pointers are aligned and we could do the move with 1 copy.

Thanks!

**Comment by felixjones** [comment]
Copying zero bytes has a bug, but I can't seem to recreate this case of "both misaligned with `size<4 && size!=0`"

Could you find a specific case and write a very minimal C reproduction of that please?

EDIT: Managed to brute force the case. 1 byte misaligned on src & dest, 1 byte size triggers it

**Comment by gwilymk** [comment]
Sorry, took me a while to get the C toolchain set up :)

```c
#include <agbabi.h>

extern void __aeabi_memcpy(void *dest, const void *src, size_t n);

int main()
{
    char buffer_char[70] = {0};

    __aeabi_memcpy(buffer_char + 5, buffer_char + 1, 3); // bug triggered if size = 0, 1 or 2

    *(unsigned int *)0x04000000 = 0x0403;

    ((unsigned short *)0x06000000)[120 + 80 * 240] = 0x001F;
    ((unsigned short *)0x06000000)[136 + 80 * 240] = 0x03E0;
    ((unsigned short *)0x06000000)[120 + 96 * 240] = 0x7C00;

    while (1)
        ;
}
```

shows the traditional 3 dots if you have 3, 4 etc, but not 0, 1 or 2 (seems I didn't test heavily enough to determine the exact lowest point sorry).

Confirmed this in my rust test as well, works with `> 2` but not `<= 2`

**Comment by felixjones** [comment]
Fixed in 8316079

The issue was that for the misaligned path the `.Lcopy4` routine always ran, which would **always** perform a copy for misaligned addresses
Also fixes the size zero copy bug

EDIT: memset and rmemcpy have the same issue, so incoming fixes for those too
